feat: Optionally sync multiple IMAP folders in `Email Account`

mtraeber opened the pull request:
If users use sieve filters in their email account and store new mail in subfolders, or manually sort new mails faster than Frappe polls the INBOX, then Frappe/ERPnext should be able to synchronize those subfolders as well. The user should be able to configure which folders are synced. The default should still be INBOX.

Created a new child table `IMAP Folder` with following fields:
Label | Fieldname | Fieldtype | Option | Field Setting
--- | --- | --- | --- | ---
Folder Name | folder_name | Data  |  | in gridview
Append To | append_to | Link | DocType | in gridview
UIDVALIDITY | uidvalidity | Data |  | hidden 
UIDNEXT | uidnext | Int |  | hidden

Doctype `Email Account` and `receive.py` code adjusted so that the emails with the changes are processed correctly and Frappe only logs in to IMAP once per sync.

Created a patch that copies the data from the old fields into the new child table with `INBOX` as default `folder_name`. This keeps existing setups working without manual changes.

Removed old fields (Append To; UIDVALIDITY; UIDNEXT) from the `Email Account` doctype, because these fields are needed for each given folder in the childtable.

All tests still are successfull.

![imap_folder_table_example](https://user-images.githubusercontent.com/57345036/111472065-e62c3d80-8729-11eb-8dfe-989c30c1836f.gif)
<!--

Some key notes before you open a PR:

 1. Select which branch should this PR be merged in?
 2. PR name follows [convention](http://karma-runner.github.io/4.0/dev/git-commit-msg.html)
 3. All tests pass locally, UI and Unit tests
 4. All business logic and validations must be on the server-side
 5. Update necessary Documentation
 6. Put `closes #XXXX` in your comment to auto-close the issue that your PR fixes


Also, if you're new here

- Documentation Guidelines => https://github.com/frappe/erpnext/wiki/Updating-Documentation

- Contribution Guide => https://github.com/frappe/frappe/blob/develop/.github/CONTRIBUTING.md

- Pull Request Checklist => https://github.com/frappe/erpnext/wiki/Pull-Request-Checklist

-->

> Docs:

docu-pr: https://github.com/frappe/frappe_docs/pull/135

mtraeber committed:
feat: sync mutliple IMAP folders in `Email Account`

When working with IMAP accounts, frappe should allow the user
to choose multiple folders to look for new mails. This helps
users to separate their frappe-related email from other
conversations. Use cases range from sieve filters in the
mail server that stuff incoming mail in various mail folders
to people manually sorting their e-mail. In both cases, we
can have different import policies for different folders, and
we can avoid importing unrelated email.

Created a new child table `IMAP Folder` with following fields:
 - Folder Name (user-modifiable)
 - Append To (user-modifiable)
 - UIDVALIDITY (hidden)
 - UIDNEXT (hidden)

Doctype `Email Account` and `receive.py` code adjusted so that
emails with the changes are processed correctly and Frappe
only logs in to the imap server once per sync.

Created a patch that copies the data from the old fields into
the new child table with `INBOX` as default `folder_name`. This
keeps existing setups working without manual changes.

The original fields
 - uidvalidity
 - uidnext
 - append_to
are still available for the pop3 setups. In IMAP, these fields
are hidden user and not used.

Added a test case in `Email Account` that validates data to make
sure a IMAP folder is provided if the use_imap is true.

Also added some code formatting changes in email_account.js to get rid
of sider checks failures that block this change

mtraeber commented:
Hi,

the last days i tried to figure out what needs to be fixed. Now only three required checks are not successful.
- "Frappe Linter / Translation Syntax Check": fails on a string in js and it's from existing and not from new code.
- "Sider": for the Sider fail, i don't know what's wrong there, because i kept the style the same way it was before there and as it is above and under my code but my code fails the check. I checked the code with PEP8 checker and it was fine there ...
- "Documentation Check / Documentation Required": i think there i need some help to success this requirement. Where i have to create the documentation for this?

Many thanks.
Manuel

hasnain2808 commented:
> Hi,
> 
> the last days i tried to figure out what needs to be fixed. Now only three required checks are not successful.
> 
> * "Frappe Linter / Translation Syntax Check": fails on a string in js and it's from existing and not from new code.

But fixing that shouldn't take long, if possible please fix them

> * "Sider": for the Sider fail, i don't know what's wrong there, because i kept the style the same way it was before there and as it is above and under my code but my code fails the check. I checked the code with PEP8 checker and it was fine there ...

Spaces are being used at some places instead of tabs. Clicking on the details link will redirect to the sider page which will point the exact issues


> * "Documentation Check / Documentation Required": i think there i need some help to success this requirement. Where i have to create the documentation for this?
> 

You can raise a PR [here](https://github.com/frappe/frappe_docs) for docs

Also please update branch and resolve conflicts

surajshetty3416 commented:
@mtraeber Any update on this?

mtraeber commented:
Hi, last week i was on vacation and will now continue on this.

I will add the docs and try to solve the conflicts

mtraeber committed:
Merge pull request #1 from frappe/develop

merge frappe/develop into LIS-Engineering/develop

mtraeber committed:
Merge branch 'develop' of https://github.com/frappe/frappe into develop

mtraeber committed:
Merge branch 'develop' of https://github.com/frappe/frappe into develop

mtraeber committed:
Merge branch 'develop' of https://github.com/frappe/frappe into develop

mtraeber committed:
Merge branch 'develop' of https://github.com/frappe/frappe into develop

mtraeber committed:
Merge branch 'frappe:develop' into develop

mtraeber committed:
Merge branch 'frappe:develop' into develop

mtraeber commented:
solved conflicts after merging the refactored code and #13412

mtraeber commented:
@leela, @surajshetty3416 or @hasnain2808  is it possible to re-open this pr? thanks

mtraeber committed:
Merge branch 'frappe:develop' into develop

mtraeber commented:
@gavindsouza thanks!

mtraeber committed:
Merge branch 'frappe:develop' into develop

mtraeber committed:
Merge branch 'frappe:develop' into develop

mtraeber committed:
Merge branch 'frappe:develop' into develop

mtraeber committed:
Merge branch 'frappe:develop' into develop

surajshetty3416 commented:
@mtraeber Please try and resolve failing checks.

Also, add few test cases to validate the feature.

Let us know if you need any help.

mtraeber committed:
Merge branch 'frappe:develop' into develop

mtraeber committed:
Merge branch 'frappe:develop' into develop

mtraeber committed:
Merge branch 'frappe:develop' into develop

mtraeber committed:
Merge branch 'frappe:develop' into develop

mtraeber committed:
Merge branch 'frappe:develop' into develop

mtraeber committed:
Merge branch 'frappe:develop' into develop

mtraeber committed:
Merge branch 'frappe:develop' into develop

mtraeber committed:
Merge branch 'frappe:develop' into develop

mtraeber committed:
Merge branch 'frappe:develop' into develop

gavindsouza commented:
@mtraeber Patch test failing due to related changes

mtraeber committed:
Merge branch 'frappe:develop' into develop

mtraeber committed:
Merge branch 'frappe:develop' into develop

mtraeber committed:
Merge branch 'frappe:develop' into develop

mtraeber committed:
Merge branch 'frappe:develop' into develop

mtraeber committed:
Merge branch 'frappe:develop' into develop

mtraeber committed:
Merge branch 'frappe:develop' into develop

mtraeber committed:
Merge branch 'frappe:develop' into develop

mtraeber committed:
Merge branch 'frappe:develop' into develop

mtraeber committed:
Merge branch 'frappe:develop' into develop

mtraeber committed:
Merge branch 'frappe:develop' into multiple_imap_folder

mtraeber committed:
Merge branch 'frappe:develop' into develop

mtraeber committed:
Merge branch 'frappe:develop' into multiple_imap_folder

mtraeber committed:
Merge branch 'frappe:develop' into develop

mtraeber committed:
Merge branch 'frappe:develop' into multiple_imap_folder

mtraeber commented:
Hi @surajshetty3416,

i think i need some help because i don't understand why three tests fail on Git, because on a fresh local install all related checks to this are successful.
As i can see the attribute `use_imap` seems to be 0 but should be 1 like it is in the testdata.

The local tests i did were: 
`bench --site <site_name> run-tests --doctype "Email Account"`

![Bildschirmfoto vom 2021-10-18 15-00-49](https://user-images.githubusercontent.com/57345036/137737025-dd5d35be-ba08-490c-aeef-35af06a149e1.png)

`bench --site <site_name> run-parallel-tests --with-coverage`

![Bildschirmfoto vom 2021-10-18 15-06-34](https://user-images.githubusercontent.com/57345036/137736967-c728c8a1-998f-4c3d-b661-026a617c4ee7.png)

Is there something I missed?

thanks

mtraeber committed:
Merge branch 'develop' into multiple_imap_folder

surajshetty3416 commented:
@mtraeber 

can you check if other tests are changing the **use_imap** setting of the email account (or make sure you have it enabled before the test begins)?

that might be the reason since `test_handle_bad_emails` and `test_imap_folder` are dependent on that setting.

mtraeber committed:
Merge branch 'frappe:develop' into develop

mtraeber committed:
Merge branch 'frappe:develop' into multiple_imap_folder

mtraeber commented:
I have just merged develop again and now the tests `test_handle_bad_emails` and `test_imap_folder` were successful

![mail_test_success](https://user-images.githubusercontent.com/57345036/139021727-cf0259dc-c38c-46dc-89e6-1a36a57e246d.png)

surajshetty3416 commented:
> I have just merged develop again and now the tests test_handle_bad_emails and test_imap_folder were successful

@mtraeber probably because tests that were affecting your test might have ran in a different instance this time. 

Just make sure to reset the email account setting as required before the test begins (in `setUp` maybe)... to avoid flaky tests

mtraeber committed:
Merge branch 'frappe:develop' into multiple_imap_folder

mtraeber committed:
Merge branch 'frappe:develop' into develop

mtraeber committed:
setting `use_imap` flag

mtraeber committed:
Merge branch 'frappe:develop' into develop

mtraeber committed:
Merge branch 'develop' into multiple_imap_folder

mtraeber committed:
added method param again after merge

mtraeber committed:
Merge branch 'frappe:develop' into develop

mtraeber committed:
Merge branch 'develop' into multiple_imap_folder

mtraeber committed:
Merge branch 'frappe:develop' into multiple_imap_folder

surajshetty3416 commented:
@mtraeber Can you update the PR with the latest develop code?

@leela can you do a final review so that this can be merged?

mtraeber committed:
Merge branch 'frappe:develop' into multiple_imap_folder

mtraeber committed:
fix: frappe-linter, all queries in email_account.py and receive.py rewritten
according to these specifications: https://frappeframework.com/docs/user/en/api/query-builder

mtraeber committed:
fix: frappe-linter, query rewritten

mtraeber committed:
fix: sider

mtraeber committed:
fix: sider

mtraeber committed:
Merge branch 'frappe:develop' into multiple_imap_folder

mtraeber committed:
Merge branch 'frappe:develop' into multiple_imap_folder

mtraeber committed:
Merge branch 'frappe:develop' into multiple_imap_folder

gavindsouza commented on frappe/email/doctype/email_account/email_account.py:
```diff
@@ -530,7 +549,11 @@ def get_unreplied_notification_emails(self):
 	def on_trash(self):
 		"""Clear communications where email account is linked"""
-		frappe.db.sql("update `tabCommunication` set email_account='' where email_account=%s", self.name)
+		Communication = frappe.qb.from_("Communication")
```
```suggestion
		Communication = frappe.qb.DocType("Communication")
```

This should fix the CI failures

mtraeber committed:
Update frappe/email/doctype/email_account/email_account.py

Co-authored-by: gavin <gavin18d@gmail.com>

leela reviewed (dismissed):
Everything looks good! Please fix the failing tests for this to get merged!

I think as a next step, we can fetch the list of mail folder's and can ask users to select from the given list.

saxenabhishek commented on frappe/email/doctype/email_account/email_account.py:
```diff
@@ -530,7 +549,11 @@ def get_unreplied_notification_emails(self):
-		frappe.db.sql("update `tabCommunication` set email_account='' where email_account=%s", self.name)
+		Communication = frappe.qb.DocType("Communication")
+		frappe.qb.update(Communication) \
+			.set(Communication.email_account == "") \
```
should fix that CI error 
```suggestion
			.set(Communication.email_account, "") \
```

mtraeber committed:
Update frappe/email/doctype/email_account/email_account.py

Co-authored-by: Abhishek Saxena <33656173+saxenabhishek@users.noreply.github.com>

mtraeber commented:
Thanks for the support!

ankush commented:
This seems be causing https://github.com/frappe/frappe/issues/15106
